pyop2/compilation.py

```diff
@@ -111,42 +108,59 @@ def get_so(self, src, extension):
         # Link into temporary file, then rename to shared library
         # atomically (avoiding races).
         tmpname = os.path.join(cachedir, "%s_p%d.so.tmp" % (basename, pid))
+        singleton = False
+
+        # To aid in compilation speed, compile only on one MPI rank per node.
+        if MPI.VERSION >= 3:
+            # If MPI3 library is available, split the communicator based on shared memory regions.
+            newcomm = self.comm.Split_type(MPI.COMM_TYPE_SHARED)
+            debug("Using MPI3 on rank ", newcomm.rank)
+            if newcomm.rank==0:
+                singleton = True
```

Comment:
I could be persuaded either way. If we restrict same disk to be same node-local temp file-system, then I don't think it's an unreasonable assumption.